tempesta_fw/apm.c

```diff
@@ -0,0 +1,1039 @@
+	ientry = entry + rbuf->rbufsz - 1;
+	for (i = 1; i < rblst->rblstmaxsz; ++i, --ientry) {
+		irbent = &rbuf->rbent[ientry % rbuf->rbufsz];
+		if (irbent->jtmistamp < jtmwstart)
+			continue;
+		rblst->rblstent[rblstsz].pcntrng = &irbent->pcntrng;
+		__tfw_apm_rngctl_copy(&rblst->rblstent[rblstsz]);
+		total_cnt += atomic64_read(&irbent->pcntrng.total_cnt);
+		rblstsz++;
+	}
+	rblst->total_cnt = total_cnt;
+	rblst->entry_cnt = entry_cnt;
+	rblst->jtmwstamp = jtmwstart;
+	rblst->rblstsz = rblstsz;
+
+	return 0;
+}
+
+/*
+ * Update the current list of ring buffer entries for the calculation
+ * of percentiles. Use optimizations to avoid the recalculation when
+ * that is possible, or to avoid a full rebuild of the current list.
+ * Build a new list when there's no other choice. Maintain current
+ * values of @entry_cnt, @total_cnt, and jtmwstamp that are used in
+ * optimizations.
+ *
+ * Return 0 if new calculation of percentiles is required.
+ * Return 1 if the percentile values don't need the recalculation.
+ */
+static int
+tfw_apm_rblst_update(TfwApmData *data)
+{
+	unsigned long jtmnow = jiffies, jtmwstart, entry_cnt;
+	TfwApmRBuf *rbuf = &data->rbuf;
+	TfwApmRBLst *rblst = &data->rblst;
+	int entry = (jtmnow / tfw_apm_jtmintrvl) % rbuf->rbufsz;
+	TfwApmRBufEnt *crbent = &rbuf->rbent[entry];
+
+	/* The start of the current time window. */
+	jtmwstart = jtmnow - (jtmnow % tfw_apm_jtmintrvl) - tfw_apm_jtmwindow;
+
+	/*
+	 * If the latest percentiles are for a different time window,
+	 * then a recalculation is in order.
+	 */
+	if (unlikely(rblst->jtmwstamp != jtmwstart))
+		return tfw_apm_rblst_buildnew(data, entry, jtmwstart);
+
+	/* The latest percentiles are for the current time window.
+	 * In some cases a recalculation is not required. In some
+	 * other cases the recalculation set up can be simpler.
+	 */
+
+	/* Nothing to do if the current entry is outdated. */
+	if (crbent->jtmistamp < jtmwstart)
+		return 1;
+
+	/* Recalculate if the current entry's data is new. */
+	if (rblst->entry_cnt == 0)
+		return tfw_apm_rblst_buildnew(data, entry, jtmwstart);
+
+	/* Nothing to do if there were no stats updates. */
+	entry_cnt = atomic64_read(&crbent->pcntrng.total_cnt);
+	if (rblst->entry_cnt == entry_cnt)
+		return 1;
+	BUG_ON(rblst->entry_cnt > entry_cnt);
+
+	/* Update the current entry's ranges data. */
+	BUG_ON(rblst->rblstent[0].pcntrng != &crbent->pcntrng);
+	tfw_apm_rngctl_copy(&rblst->rblstent[0], crbent);
+
+	/* Update the counts incrementally. */
+	rblst->total_cnt += entry_cnt - rblst->entry_cnt;
+	rblst->entry_cnt = entry_cnt;
+
+	return 0;
+}
+
+/*
+ * Calculate the latest percentiles if necessary.
+ *
+ * Return 0 if potentially new percentile values were calculated.
+ * Return 1 if the percentile values didn't need the recalculation.
+ */
+static int
+__tfw_apm_calc(TfwApmData *data, TfwApmPrcntl *prcntl)
+{
+	TfwApmRBLst *rblst = &data->rblst;
+	int i, rblstsz = rblst->rblstsz;
+
+	if (tfw_apm_rblst_update(data))
+		return 1;
```

Comment:
I believe we don't need `TfwApmRBLst` at all and can make all the calculations over the initial ring buffer, so the function should be eliminated.